### PR TITLE
[GOBBLIN-1841] Move disabling of current live instances to the GobblinClusterManager startup

### PR DESCRIPTION
**Pull request by Peiyingy**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1841] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1841

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

The method `disableLiveHelixInstances()` originally exists in `GobblinYarnAppLauncher` to disable pre-existing live instances in a Helix cluster. However, there can be an edge case when the `GobblinYarnAppLauncher` is bypassed due to a lost Application Master, like an automated replacement of the Application Master by Yarn, which will bypass the Azkaban Yarn App Launcher. In such cases, the `disableLiveHelixInstances()` will be not invoked.
Thus, we moved this method to `GobblinClusterManager`, which is at the Application Master level, and invokes it at the start stage. In this way, we can guarantee that it won't miss the edge case, and each time these pre-existing live instances can be disabled.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Screenshot 2023-06-21 at 10 21 26 AM

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java [hunk 1]
```diff
@@ -280,6 +281,7 @@ public synchronized void start() {
 
     this.eventBus.register(this);
     this.multiManager.connect();
```
Comment (homatthew) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java:283]: The multimanager only connects the job cluster helix manager if we are in dedicated mode. This class is also shared between YARN and non-YARN apps. 

But we are moving the previous logic from a yarn specific class. Maybe it makes more sense to have the disableLiveHelixInstances method as part of the GobblinApplicationMaster which extends this class but is yarn specific

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java [hunk 2]
```diff
@@ -280,6 +281,7 @@ public synchronized void start() {
...
+    disableLiveHelixInstances();
```
Comment (homatthew) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java:284]: Seems like this should only be enabled if we are non standalone mode

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java [hunk 3]
```diff
@@ -504,6 +506,18 @@ private static void printUsage(Options options) {
     formatter.printHelp(GobblinClusterManager.class.getSimpleName(), options);
   }
 
+  public void disableLiveHelixInstances() {
+    HelixManager helixManager = this.multiManager.getJobClusterHelixManager();
```
[the remark this answers is not in the document]
Comment (homatthew) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java:519]: Upon further investigation, this job cluster manager should be connected elsewhere. But I see another issue where you are fetching all live instances (include itself!). Which means that when the AM calls this method, it will disable itself.

Please take a look at the below code https://github.com/apache/gobblin/blob/5af6bca57df909e44b995e5b2d667c70e0399877/gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnAutoScalingManager.java#L187C1-L196C6

This code fetches all live participants that are task runners. So we should only disable those instances.

And then we can also rename this method to `disableTaskRunnersFromPreviousExecutions`.

**Review by homatthew — CHANGES_REQUESTED** [review]

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java [hunk 4]
```diff
@@ -504,6 +516,7 @@ private static void printUsage(Options options) {
     formatter.printHelp(GobblinClusterManager.class.getSimpleName(), options);
   }
 
+
```
Comment (homatthew) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterManager.java:519]: Remove random whitespace. This will probably fail the linter

File: gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnAutoScalingManager.java [hunk 5]
```diff
@@ -272,7 +260,7 @@ void runInternal() {
       }
       // Find all participants appearing in this cluster. Note that Helix instances can contain cluster-manager
       // and potentially replanner-instance.
-      Set<String> allParticipants = getParticipants(HELIX_YARN_INSTANCE_NAME_PREFIX);
+      Set<String> allParticipants = HelixUtils.getParticipants(helixDataAccessor,HELIX_YARN_INSTANCE_NAME_PREFIX);
```
Comment (homatthew) [inline gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnAutoScalingManager.java:263]: Add whitespace after the comma. This will probably fail the linter

File: gobblin-yarn/src/test/java/org/apache/gobblin/yarn/GobblinApplicationMasterTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    PropertyKey.Builder mockBuilder = Mockito.mock(PropertyKey.Builder.class);
+    when(mockAccessor.keyBuilder()).thenReturn(mockBuilder);
+
+    PropertyKey mockLiveInstancesKey = Mockito.mock(PropertyKey.class);
+    when(mockBuilder.liveInstances()).thenReturn(mockLiveInstancesKey);
+
+    int instanceCount = 3;
+    Map<String, HelixProperty> mockChildValuesMap = new HashMap<>();
+    for (int i = 0; i < instanceCount; i++) {
+      mockChildValuesMap.put("GobblinYarnTaskRunner_TestInstance_" + i, Mockito.mock(HelixProperty.class));
```
Comment (homatthew) [inline gobblin-yarn/src/test/java/org/apache/gobblin/yarn/GobblinApplicationMasterTest.java:47]: We should also be adding non `GobblinYarnTaskRunner`s as part of the cluster. E.g. `GobblinClusterManager`. Because we'd only want to disable those that start with `GobblinYarnTaskRunner`. 

And then in your mockito verify, you should verify we never call disable on `GobblinClusterManager`

**Review by homatthew — CHANGES_REQUESTED** [review]

File: gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinApplicationMaster.java [hunk 7]
```diff
@@ -135,6 +141,26 @@ protected MultiTypeMessageHandlerFactory getUserDefinedMessageHandlerFactory() {
...
+
+  public static void disableTaskRunnersFromPreviousExecutions(GobblinHelixMultiManager multiManager) {
+    HelixManager helixManager = multiManager.getJobClusterHelixManager();
+    HelixDataAccessor helixDataAccessor = helixManager.getHelixDataAccessor();
+    String clusterName = helixManager.getClusterName();
+    HelixAdmin helixAdmin = helixManager.getClusterManagmentTool();
+    Set<String> taskRunners = HelixUtils.getParticipants(helixDataAccessor,
+        GobblinYarnTaskRunner.HELIX_YARN_INSTANCE_NAME_PREFIX);
+    LOGGER.warn("Found {} task runners in the cluster.", taskRunners.size());
+    for (String taskRunner: taskRunners) {
```
Comment (homatthew) [inline gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinApplicationMaster.java:158]: whitespace before the `:`. But this will probably be caught by the linter

Comment (homatthew) [inline gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinApplicationMaster.java:158]: whitespace before the `:`. But this will probably be caught by the linter

**Review by homatthew — CHANGES_REQUESTED** [review]

**Review by homatthew — APPROVED** [review]
LGTM! Great work 🚢

File: gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnAppLauncher.java [hunk 8]
```diff
@@ -540,26 +537,6 @@ void connectHelixManager() {
...
 
-  /**
-   * A method to disable pre-existing live instances in a Helix cluster. This can happen when a previous Yarn application
-   * leaves behind orphaned Yarn worker processes. Since Helix does not provide an API to drop a live instance, we use
-   * the disable instance API to fence off these orphaned instances and prevent them from becoming participants in the
-   * new cluster.
-   *
-   * NOTE: this is a workaround for an existing YARN bug. Once YARN has a fix to guarantee container kills on application
-   * completion, this method should be removed.
-   */
```
Comment (homatthew) [inline gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnAppLauncher.java:551]: Can you add this comment back to the new impl

**Review by homatthew — CHANGES_REQUESTED** [review]
Can you make sure to add back the comment from the previous impl

**Review by homatthew — APPROVED** [review]

**Review by ZihanLi58 — APPROVED** [review]
+1
